src/Client/ConnectionEstablisher.h

```diff
@@ -24,7 +24,9 @@ class ConnectionEstablisher
                           const QualifiedTableName * table_to_check = nullptr);
 
     /// Establish connection and save it in result, write possible exception message in fail_message.
-    void run(TryResult & result, std::string & fail_message);
+    /// The connection is returned from the pool, it can be stale. Use force_connected flag
+    /// to ensure that connection is working one
+    void run(TryResult & result, std::string & fail_message, bool force_connected = false);
```

Comment:
it looks a little concerning that we want it to be false sometimes. but let's at least make the default choice to be true?

Reply:
I thought similarly before, but it turns out there are a couple of points to consider (in sequence it happened to me)

(1) number of tests fails - [Integration Tests (asan, Old Analyzer) [5/6]](https://s3.amazonaws.com/clickhouse-test-reports/67389/82c06920e35238e6d411396e1d67d9fc603d5a04/integration_tests__asan__old_analyzer__[5_6].html) and [Integration Tests (asan, Old Analyzer) [1/6]](https://s3.amazonaws.com/clickhouse-test-reports/67389/82c06920e35238e6d411396e1d67d9fc603d5a04/integration_tests__asan__old_analyzer__[1_6].html). And it seems, it'd be quite an endeavor to reconsider their behavior

(2) another thing, to consider:
`force_connected` implies sending `Ping` packet to another peer and if it fails - reestablishing the connection. `ping-pong` round trip can be unnecessary in case of connection is still alive.  It can be the case when the caller does retries, i.e. it handles stale connections by retrying, - `ConnectionPoolWithFailover` works this way. With parallel replicas, we don't do retries - just use replicas which are available from the first attempt to connect.